### ISSUE TITLE
[Feature] Real-Time Generation-Based Evaluation for SFT (Supervised Fine-Tuning)

### ISSUE THREAD

**Issue by nelish01** [meta-pytorch/forge#665]
### Context/Motivation

TorchForge currently supports only loss-based evaluation metrics during SFT training. As shown in our internal investigations, this can miss critical issues in output structure or correctness even when loss improves. We propose a framework for real-time generation-based evaluation within the SFT training pipeline, including a flexible metric registry and domain-specific evaluation packs, to catch model collapse, ensure semantic and structural output quality, and enable rapid extensibility for new tasks/domains.

### Pseudo-code + acceptance criteria [Optional]

# In SFT pipeline

if step % gen_eval_every_n_steps == 0:
    await self.run_generation_eval_and_record()

run_generation_eval_and_record will:
1. Unwrap FSDP to expose base transformer (if needed)
2. Wrap with TitanGenerateAdapter granting .generate() interface
3. Load packed/unpacked raw eval data, extract prompts/refs
4. Generate outputs and compare against references via metric packs
5. Log metrics, trigger early stop if required

Acceptance criteria:
- Generation-based evaluation can be enabled/disabled via YAML config and runs at configurable intervals.
- Any metric compatible with registry can be plugged in with a decorator.
- Domain- and task-specific metric packs are auto-expanded in config.
- Metrics and results are logged and actionable (for reporting, early stop, or debugging).
- All components operate on multi-GPU/FSDP-compatible setups.

See long-form investigation + data below for migration details.

**Comment by nelish01** [comment]
Here is the detailed experimentation results for your reference

https://github.com/nelish01/torchforge_test/blob/main/apps/sft/fine-tuning_use_case/instamart_finetuning/TORCHFORGE_GENERATION_EVAL_TECHNICAL_DOC.md

**Comment by felipemello1** [comment]
Hey @nelish01 , thanks for sharing it! We have in our roadmap to add generation eval to SFT script. We don't have bandwidth to look at this atm, but we will probably get to it in a few weeks and will get back to this thread. We would love to have someone contribute with a PR, but we are currently refactoring our trainer, so its not good timing. Is it ok if we come back to this some time after the holidays?

**Comment by nelish01** [comment]
Hi @felipemello1, thanks for the update! I'm excited to hear this and looking forward to collaborating with you on this!

Happy holidays and lets connect after this :)

**Comment by kimbriancanavan** [comment]
Hey @felipemello1, @nelish01! I'm curious where this and the trainer refactors that @felipemello1 mentioned are up to? I'd be interested in helping out on this if I can.

**Comment by felipemello1** [comment]
hey @kimbriancanavan , thanks for the interest! I believe that @nelish01 was going to work on an RFC proposing it. Is it still the case @nelish01 ?
